src/compression/mod.rs

```diff
@@ -0,0 +1,394 @@
+use std::{fmt, str::FromStr};
+
+#[cfg(feature = "zlib-compression")]
+use flate2::{
+    write::{ZlibDecoder, ZlibEncoder},
+    Compression,
+};
+
+#[cfg(any(
+    feature = "zstd-compression",
+    feature = "zlib-compression",
+    feature = "snappy-compression"
+))]
+use std::io::prelude::*;
+
+#[cfg(any(
+    feature = "zstd-compression",
+    feature = "zlib-compression",
+    feature = "snappy-compression"
+))]
+use crate::error::Result;
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg(any(
+    feature = "zstd-compression",
+    feature = "zlib-compression",
+    feature = "snappy-compression"
+))]
+pub(crate) enum CompressorId {
+    Noop = 0,
+    #[cfg(feature = "snappy-compression")]
+    Snappy = 1,
+    #[cfg(feature = "zlib-compression")]
+    Zlib = 2,
+    #[cfg(feature = "zstd-compression")]
+    Zstd = 3,
+}
+
+#[cfg(any(
+    feature = "zstd-compression",
+    feature = "zlib-compression",
+    feature = "snappy-compression"
+))]
+impl CompressorId {
+    pub(crate) fn from_u8(id: u8) -> Result<Self> {
+        match id {
+            0 => Ok(CompressorId::Noop),
+            #[cfg(feature = "snappy-compression")]
+            1 => Ok(CompressorId::Snappy),
+            #[cfg(feature = "zlib-compression")]
+            2 => Ok(CompressorId::Zlib),
+            #[cfg(feature = "zstd-compression")]
+            3 => Ok(CompressorId::Zstd),
+            other => Err(ErrorKind::InvalidResponse {
+                message: format!("Invalid compressor id: {}", other),
+            }
+            .into()),
+        }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[non_exhaustive]
+#[cfg(any(
+    feature = "zstd-compression",
+    feature = "zlib-compression",
+    feature = "snappy-compression"
+))]
+/// Enum representing supported compressor algorithms.
+/// Used for compressing messages sent to the server.
+/// For compressors that take a level, use None to indicate the default level.
+/// Requires one of zstd-compression, zlib-compression, or snappy-compression
+/// feature flags.
+pub enum Compressor {
+    /// Zstd compressor.  Level 0 is the default.  Requires Rust version 1.54.
+    /// See Zstd [manual](http://facebook.github.io/zstd/zstd_manual.html) for more information
+    #[cfg(feature = "zstd-compression")]
+    Zstd { level: Option<i32> },
+    /// Zlib compressor.  Levels -1 through 9 are valid.
+    /// Level -1 indicates default level.
+    /// Level 0 indicates no compression.
+    /// Higher level indicates more compression (and slower).
+    #[cfg(feature = "zlib-compression")]
+    Zlib { level: Option<i32> },
+    /// Snappy compressor.
+    #[cfg(feature = "snappy-compression")]
+    Snappy,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[non_exhaustive]
+#[cfg(not(any(
+    feature = "zstd-compression",
+    feature = "zlib-compression",
+    feature = "snappy-compression"
+)))]
+pub(crate) enum Compressor {}
+
+impl FromStr for Compressor {
```

Comment:
Instead of implementing `FromStr` (which will be present in the public API), can we instead make this a `pub(crate)` method on `Compressor`? That way, users won't accidentally miss the level option or misspell the compressor name.